libraries/GCS_MAVLink/GCS_Common.cpp

```diff
@@ -431,44 +484,124 @@ void GCS_MAVLINK::send_ahrs3()
 #endif
 }
 
-/*
-  handle a MISSION_REQUEST_LIST mavlink packet
- */
-void GCS_MAVLINK::handle_mission_request_list(AP_Mission &mission, mavlink_message_t *msg)
+MissionItemProtocol *GCS::get_prot_for_mission_type(const MAV_MISSION_TYPE mission_type) const
+{
+    switch (mission_type) {
+    case MAV_MISSION_TYPE_MISSION:
+        return _waypoint;
+    case MAV_MISSION_TYPE_RALLY:
+        return _rally;
+    case MAV_MISSION_TYPE_FENCE:
+        return _fence;
+    default:
+        return nullptr;
+    }
+}
+
+// handle a request for the number of items we have stored for a mission type:
+void GCS_MAVLINK::handle_mission_request_list(const mavlink_message_t *msg)
 {
     // decode
     mavlink_mission_request_list_t packet;
     mavlink_msg_mission_request_list_decode(msg, &packet);
 
-    // reply with number of commands in the mission.  The GCS will then request each command separately
-    mavlink_msg_mission_count_send(chan,msg->sysid, msg->compid, mission.num_commands(),
-                                   MAV_MISSION_TYPE_MISSION);
+    MissionItemProtocol *prot = gcs().get_prot_for_mission_type((MAV_MISSION_TYPE)packet.mission_type);
+    if (prot == nullptr) {
+        mavlink_msg_mission_ack_send(chan,
```

Comment:
How do you know there is buffer space to send this?

Reply:
We don't.  We should, but we don't.

We don't on master either.

We *will* fix this.  But not as part of this PR :-)